src/qibolab/instruments/qm/acquisition.py

```diff
@@ -175,6 +197,11 @@ class ShotsAcquisition(Acquisition):
     """Stream to collect multiple shots."""
 
     def __post_init__(self):
+        if self.threshold is None or self.angle is None:
+            raise ValueError(
+                "Cannot use ``AcquisitionType.DISCRIMINATION`` "
+                "if threshold and angle are not given."
+            )
         self.cos = np.cos(self.angle)
```

Comment:
The only argument is to have a more readable error, but I understand that this can blow up the code so I removed in 555b12fa4c4dde3723b799211dc04de9fea94c4c.

That being said, I am not sure what is the correct approach when it comes to custom errors. For this case in particular, assume the following situation that can easily happen: the user provides a runcard without specifying neither threshold nor angle and asks for `DISCRIMINATION`. The driver will query `Qubit` for threshold and angle and get the default values which are `None` and `0` respectively. Therefore there will be no error in `__post_init__` (since cos and sin of `0` can be calculated) and the first error will appear in
https://github.com/qiboteam/qibolab/blob/6f8596621078cde059785f053ce2332f36c1116e/src/qibolab/instruments/qm/acquisition.py#L221
because `self.threshold is None`. If QUA provides a readable error there, fine, if no the user will have not idea what happened (ie. that he cannot ask for `DISCRIMINATION` if he does not provide thresholds in the runcard).

Reply:
> That being said, I am not sure what is the correct approach when it comes to custom errors.

I believe this case is similar to the relative imports we discussed with @hay-k: it depends.
In general, in Python the attempt is to duck type functions: until they work, they work. If they fail, it should be because some operation is impossible on the given value, then that operation is reported.

However, as you pointed out, this could generate deeply nested errors, when you were able to tell it would have caused an error much before.
My advice would be to often resist the temptation to catch errors too much in advance, because it will make your code quite tangled.

We could start to validate the user input, to catch some mismatch in user input immediately, and for this task `pydantic` would be really ideal. However, Qibolab has quite experienced users (like Qibocal), so I'd consider this a second-order problem for some more time.